### PR TITLE
Upgrade to Webpack 4

### PR DESCRIPTION
**Pull request by pascalw**
Upgrade was quite straightforward.

Performance should be improved quite a bit. The example project compiles in ~ 3.6 seconds vs 7s.

I did have to switch some loaders/plugins to beta releases, so YMMV. Let's try running some projects on this branch before we merge.

I snug in one technically maybe-not-required-but-nice change. I had to use `babel-loader` 8, which is only compatible with babel 7. `babel-preset-env` on Babel 7 is much better and doesn't require `babel-polyfill`, so I removed that.

### CODE REVIEW

File: project_template/config/karma.js [hunk 1]
```diff
@@ -1,8 +1,4 @@
 const webpackConfig = require("../webpack.config.js");
-const commonsChunkPluginIndex = webpackConfig.plugins.findIndex(
```
Comment (pascalw) [inline project_template/config/karma.js]: @ruisalgado do you remember why this was needed earlier?

Comment (ruisalgado) [reply project_template/config/karma.js]: @pascalw nope. Seems like we're trying to exclude some plugins from being loaded by karma but... no idea why

File: project_template/package.json [hunk 2]
```diff
@@ -53,38 +52,39 @@
     "eslint-plugin-compat": "^1.0.4",
     "eslint-plugin-prettier": "^2.1.2",
     "eslint-plugin-react": "^7.1.0",
-    "extract-text-webpack-plugin": "^2.1.0",
-    "file-loader": "^0.11.2",
-    "html-webpack-plugin": "^2.28.0",
+    "extract-text-webpack-plugin": "4.0.0-beta.0",
```
Comment (ruisalgado) [inline project_template/package.json]: we need to keep an eye on these beta dependencies

File: project_template/webpack.config.js [hunk 3]
```diff
@@ -29,32 +29,31 @@ const plugins = [
...
   new Clean(["dist"], { verbose: false, exclude: [".keep"] }),
-  new webpack.optimize.CommonsChunkPlugin({
-    name: "vendor"
-  }),
   extractShellCss,
   extractOtherCss,
   new SpriteLoaderPlugin()
 ];
 
+let optimization = {
```
Comment (ruisalgado) [inline project_template/webpack.config.js]: I suggest making this immutable:

```javascript
const optimization: {
  splitChunks: {
    name: "vendor"
  },
  minimize: isProd,
  namedModules: !isProd,
  noEmitOnErrors: !isProd
}
```

File: project_template/package.json (hunk 2 above)
Comment (pascalw) [reply project_template/package.json]: Yep. I was hoping these plugins would be updated to final by the time we merge this.

File: project_template/webpack.config.js (hunk 3 above)
Comment (pascalw) [reply project_template/webpack.config.js]: Good point, done!

File: project_template/webpack.config.js [hunk 4]
```diff
@@ -29,31 +29,26 @@ const plugins = [
...
-} else {
-  if (hotReload) {
-    plugins.push(
-      // enable HMR globally
-      new webpack.HotModuleReplacementPlugin()
-    );
-  }
+const optimization = {
+  splitChunks: {
+    name: "vendor"
```
Comment (ruisalgado) [inline project_template/webpack.config.js]: @pascalw in my experience using the default webpack4 heuristics produces better results. This is what I get with the setup in this PR:

screen shot 2018-05-10 at 17 52 44

It seems like all the vendor stuff is still in the main `app.js`, and a separate chunk is created for the `vendors` entrypoint which ends up having duplicate stuff.

Running _without_ the vendors entrypoint and with the following optimization produces much better results:

```javascript
  splitChunks: {
    chunks: "all"
  }
```

screen shot 2018-05-10 at 17 52 59

As far as I can tell webpack 4 already expects to produce a vendors chunk per entrypoint (or multiple, depending on the size of the dependencies).

Comment (pascalw) [reply project_template/webpack.config.js]: That's cool! I'll update this PR to let Webpack do it's job :-)

File: project_template/webpack.config.js [hunk 5]
```diff
@@ -36,7 +36,7 @@ const plugins = [
 
 const optimization = {
   splitChunks: {
```
Comment (ruisalgado) [inline project_template/webpack.config.js]: @pascalw upon further exploration its best to disable chunking when not in production, as it causes some issues with Karma and unnecessary delays in test execution: `splitChunks: isProd && { chunks: "all" }`.

Issue I witnessed (with chunking on) was Karma silently failing to load tests that import code that makes use of the `h` pragma for some reason. Also saw webpack producing chunks for the Karma entrypoint which is totally unnecessary.

Comment (pascalw) [reply project_template/webpack.config.js]: Ah gotcha. This is probably why we had this chunks related config in Karma config before.

**Comment by pascalw** [comment]
@ruisalgado I've updated the PR to only split chunks in prod mode. Have you encountered any other issues or is this good to go?

**Comment by ruisalgado** [comment]
@pascalw haven't bumped into further issues so I'd be ok merging this

**Comment by pascalw** [comment]
Sorry for the delay!

File: project_template/package.json [hunk 6]
```diff
@@ -36,13 +36,12 @@
   },
   "devDependencies": {
     "autoprefixer": "^7.1.1",
-    "babel-core": "^6.24.1",
+    "@babel/core": "^7.0.0-beta.44",
```
Comment (pascalw) [inline project_template/package.json]: Unfortunately all of these beta modules are still not released.
They seem to be pretty stable, but for the time being I'll hold off creating a new Maji release.

**Pull request merged**